### PR TITLE
buildTargetScalaTestClasses implementation 

### PR DESCRIPTION
**Pull request by Abrams27**
my implementation of `buildTargetScalaTestClasses` method

### CODE REVIEW

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java [hunk 1]
```diff
@@ -39,19 +54,65 @@ public ScalaBuildServerService(BazelData bazelData, BazelRunner bazelRunner) {
...
+                    attribute, SCALA_TEST_SRCS_CLASSES_ATTRIBUTE_NAME))
+        .map(Attribute::getStringListValueList)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+  }
+
+  private boolean doesAttributesContainTestClass(Build.Rule rule) {
+    return TargetsUtils.doesRuleAttributesContain(rule, SCALA_TEST_MAIN_CLASSES_ATTRIBUTE_NAME)
+        || TargetsUtils.doesRuleAttributesContain(rule, SCALA_TEST_SRCS_CLASSES_ATTRIBUTE_NAME);
   }
```
Comment (magda-aug) [inline src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:116]: Shouldn't we check name of the rule in order to find test classes? ('scala_test' probably?)

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/resolvers/TargetsUtils.java [hunk 2]
```diff
@@ -9,4 +10,18 @@
   public static List<String> getTargetsUris(List<BuildTargetIdentifier> targets) {
     return targets.stream().map(BuildTargetIdentifier::getUri).collect(Collectors.toList());
   }
+
+  public static boolean doesRuleAttributesContain(Build.Rule rule, String attributeName) {
```
[the remark this answers is not in the document]
Comment (Abrams27) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/resolvers/TargetsUtils.java:14]: agree

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java (hunk 1 above)
Comment (Abrams27) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:116]: right! I forgot to check it

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java [hunk 3]
```diff
@@ -1,32 +1,48 @@
...
 
 public class ScalaBuildServerService {
 
   private static final String SCALA_COMPILER_OPTIONS_NAME = "scalacopts";
   private static final List<String> SCALA_LANGUAGES_IDS =
       ImmutableList.of(Constants.SCALAC, Constants.JAVAC);
 
+  private static final String SCALA_TEST_RULE_NAME = "scala_test";
+  private static final String SCALA_TEST_MAIN_CLASSES_ATTRIBUTE_NAME = "main_class";
+  private static final String SCALA_TEST_SRCS_CLASSES_ATTRIBUTE_NAME = "srcs";
```
Comment (andrefmrocha) [inline src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:38]: These attributes are generic to all `_test` and `_binary`. Maybe this should live in the `Constants`?

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java [hunk 4]
```diff
@@ -39,19 +55,64 @@ public ScalaBuildServerService(BazelData bazelData, BazelRunner bazelRunner) {
...
+    List<String> classes = getTestClasses(rule);
+
+    return new ScalaTestClassesItem(target, classes);
+  }
+
+  private List<String> getTestClasses(Build.Rule rule) {
+    List<String> mainClasses = getTestMainClasses(rule);
+    List<String> srcsClasses = getTestSrcsClasses(rule);
+
+    return Stream.concat(mainClasses.stream(), srcsClasses.stream()).collect(Collectors.toList());
```
Comment (andrefmrocha) [inline src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: `main_class` is the defined name to the class where the main is, while `srcs` has the path to all the source files in a given rule. I'm not sure what you're trying to do here, but it doesn't seem correct. Can you pls explain?

Comment (andrefmrocha) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: More info on those values here: https://docs.bazel.build/versions/master/be/java.html#java_test

This is the docs for java, but all languages must abide by these rules

**Review by andrefmrocha — CHANGES_REQUESTED** [review]
I have some doubts as to what is being done here. Also, we probably shouldn't merge this without tests right?

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java (hunk 4 above)
Comment (Abrams27) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: I checked on more time and I agree that I misunderstood how `srcs` attribute works. But, now I found out that I should handle `test_class` attribute as well.

Comment (andrefmrocha) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: I wanted to check what would happen if both attributes were present and how to approach this and ended up finding out that `rules_scala` doesn't actually support `test_class`.

Comment (Abrams27) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: oh great, soo should I remove `srcsClasses` and connected logic and use only `main_class` attribute?

Comment (andrefmrocha) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: I do believe so!

Comment (andrefmrocha) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: Whenever they end up supporting more stuff that makes need to support `test_class`, we'll look into it. If that even happens :smile:

Comment (Abrams27) [reply src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java:90]: cool

**Review by andrefmrocha — APPROVED** [review]
Lgtm :rocket: !

**Comment by andrefmrocha** [comment]
Are we handling tests in this PR or separately>

**Comment by Abrams27** [comment]
> Are we handling tests in this PR or separately?

I will introduce tests in this PR, but firstly need to merge https://github.com/JetBrains/bazel-bsp/pull/32

**Review by agluszak — APPROVED** [review]

**Review by gerardd33 — APPROVED** [review]
Looks good!

File: src/main/java/org/jetbrains/bsp/bazel/server/bsp/impl/BuildServerImpl.java [hunk 5]
```diff
@@ -110,4 +110,10 @@ public void onBuildExit() {
     return serverRequestHelpers.executeCommand(
         () -> buildServerService.buildTargetCleanCache(cleanCacheParams));
   }
+
+  @Override
+  public CompletableFuture<Object> workspaceReload() {
+    // TODO
+    return null;
```
Comment (andrefmrocha) [inline src/main/java/org/jetbrains/bsp/bazel/server/bsp/impl/BuildServerImpl.java:117]: It's probably best to return a [Completed Future](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#completedFuture-U-) here

File: sample-repo/example/ExampleTest.scala [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import dep.Dep
+
+object ExampleTest {
+  def test(): Unit = {
```
Comment (andrefmrocha) [inline sample-repo/example/ExampleTest.scala:4]: To keep this consistent, this object should have a main method, since this is required by main_class

**Review by andrefmrocha — CHANGES_REQUESTED** [review]
Just some minor nitpicks

**Review by andrefmrocha — APPROVED** [review]
Lgtm :rocket: !

File: src/test/java/org/jetbrains/bsp/bazel/BazelBspServerTest.java [hunk 7]
```diff
@@ -15,6 +15,9 @@
 
   private static final Logger LOGGER = LogManager.getLogger(BazelBspServerTest.class);
 
+  private static final Integer SUCCES_EXIT_CODE = 0;
```
Comment (jastice) [inline src/test/java/org/jetbrains/bsp/bazel/BazelBspServerTest.java:18]: typo: success

**Review by jastice — APPROVED** [review]
LGTM!

**Pull request merged**